File: runme/mcgen/local/datan/datsrc/minasy.c
#include <stdlib.h>
#include <limits.h>
#include <math.h>
#include "datsrc.h"

void LIBFUNC minasy_(double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), integer const P_T *n,
             integer const P_T *nred, integer P_T *list, double const P_T *x0,
             double const P_T *cx, double const P_T *fcont, double P_T *dxplus,
             double P_T *dxmins, double P_T *xr, integer P_T *nstep)
{
    /* System generated locals */
    integer xr_dim1, xr_offset, cx_dim1, cx_offset, i__1, i__2;
    double d__1, d__2;

    /* Local variables */
    integer isign, ivar, nred1, i, nstepl;
    double del, sav, dxmin, xivar, r1, xbig, epsi, xsmall, signum;
    logical lencls;
    /*double x[100], dx[100];*/
    double *dx, *x;
    unsigned long amem;

    if ((amem = sizeof(double) * *n) > MAXALLOC) {
        MEMERROR("minasy_: x (too large)");
        goto f_0;
    }
    x = (double *)MEMALLOC(amem);
    if (!x) {
        MEMERROR("minasy_: x");
        goto f_0;
    }
    if ((amem = sizeof(double) * max(*n,*nred)) > MAXALLOC) {
        MEMERROR("minasy_: dx (too large)");
        goto f_1;
    }
    dx = (double *)MEMALLOC(amem);
    if (!dx) {
        MEMERROR("minasy_: dx");
        goto f_1;
    }

    /* Parameter adjustments */
    --x0;
    --list;
    xr_dim1 = *nred;
    xr_offset = xr_dim1 + 1;
    xr -= xr_offset;
    --dxmins;
    --dxplus;
    cx_dim1 = *nred;
    cx_offset = cx_dim1 + 1;
    cx -= cx_offset;

    /* Function Body */
    if (*nstep <= 0) {
        *nstep = 1000;
    }
    if (*n == *nred) {
        i__1 = *n;
        for (i = 1; i <= i__1; ++i) {
            list[i] = 1;
        }
    }
    dxmin = 1e20;
    i__1 = *nred;
    for (i = 1; i <= i__1; ++i) {
        dx[i - 1] = sqrt(cx[i + i * cx_dim1]);
/* Computing MIN */
        d__1 = dxmin, d__2 = dx[i - 1];
        dxmin = min(d__1,d__2);
    }
    dxmin = .01 * dxmin;
    nred1 = *nred - 1;
    i__1 = *n;
    for (ivar = 1; ivar <= i__1; ++ivar) {
        if (list[ivar] == 0) {
            goto L60;
        }
/* fix variable IVAR */
        list[ivar] = 0;
        sav = x0[ivar];
        for (isign = -1; isign <= 1; isign += 2) {
            signum = (double) isign;
            del = dx[ivar - 1];
            lencls = FALSE_;
/* set XSMALL to x at minimum position */
            xsmall = x0[ivar];
            xivar = x0[ivar] + signum * del;
            i__2 = *nstep;
            for (i = 1; i <= i__2; ++i) {
                nstepl = *nstep;
                mtxcpv_(&x0[1], x, n);
                x[ivar - 1] = xivar;
                epsi = 0.;
                r1 = dxmin;
                minsim_(x, n, &nred1, &list[1], userfn, &r1, &epsi, &nstepl,
                        &xr[xr_offset]);
                if (nstepl < 1) {
                    *nstep = -1;
                    goto L70;
                }
/* test for convergence */
                if ((d__1 = r1 - *fcont, abs(d__1)) < .001) {
                    if (isign < 0) {
                        dxmins[ivar] = x0[ivar] - x[ivar - 1];
                    } else {
                        dxplus[ivar] = x[ivar - 1] - x0[ivar];
                    }
                    goto L40;
                } else {
                    if (! lencls) {
/* zero was not yet enclosed in last step */
                        if (r1 > *fcont) {
/* zero is now enclosed, perform first interval halving */
                            lencls = TRUE_;
                            xbig = x[ivar - 1];
                            x[ivar - 1] = (xbig + xsmall) * .5;
                        } else {
/* zero not enclosed, widen range */
                            del *= 2.;
                            x[ivar - 1] = xsmall + signum * del;
                        }
                    } else {
/* continue interval halving */
                        if (r1 > *fcont) {
                            xbig = x[ivar - 1];
                        } else {
                            xsmall = x[ivar - 1];
                        }
                        x[ivar - 1] = (xbig + xsmall) * .5;
                    }
                    xivar = x[ivar - 1];
                }
            }
            *nstep = -2;
            goto L70;
L40:
            ;
        }
/* unfix variable IVAR */
        list[ivar] = 1;
L60:
        ;
    }
L70:
    MEMFREE((void *)dx);
f_1:
    MEMFREE((void *)x);
f_0:
    return;
} /* minasy_ */

File: runme/mcgen/local/datan/datsrc/minasy.f
      SUBROUTINE MINASY(USERFN,N,NRED,LIST,X0,CX,FCONT,DXPLUS,
     +DXMINS,XR,NSTEP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=100)
      DIMENSION X(NMAX),DX(NMAX),LIST(N),XR(NRED,NRED+1)
      DIMENSION CX(NRED,NRED),DXPLUS(NRED),DXMINS(NRED)
      DIMENSION X0(N)
      PARAMETER(MAXSTP=1000)
      PARAMETER(EPS=1.D-3,EPSILN=1.D-4,TT=1.D-6,FACT=1.5D0)
      PARAMETER(ZERO=0.0D0,ONE=1.D0,HALF=0.5D0,TWO=2.D0,BIG=1.D20)
      LOGICAL LENCLS
      EXTERNAL USERFN
      IF(NSTEP.LE.0) NSTEP=MAXSTP
      IF(N.EQ.NRED) THEN
        DO 10 I=1,N
          LIST(I)=1
   10   CONTINUE
      END IF
      DXMIN=BIG
      DO 20 I=1,NRED
        DX(I)=SQRT(CX(I,I))
        DXMIN=MIN(DXMIN,DX(I))
   20 CONTINUE
      DXMIN=DBLE(.01)*DXMIN
      NRED1=NRED-1
      DO 60 IVAR=1,N
        IF(LIST(IVAR).EQ.0) GO TO 60
C fix variable IVAR
        LIST(IVAR)=0
        SAV=X0(IVAR)
        DO 50 ISIGN=-1,1,2
          SIGNUM=DBLE(ISIGN)
          DEL=DX(IVAR)
          LENCLS=.FALSE.
C set XSMALL to x at minimum position
          XSMALL=X0(IVAR)
          XIVAR=X0(IVAR)+SIGNUM*DEL
          DO 30 I=1,NSTEP
            NSTEPL=NSTEP
            CALL MTXCPV(X0,X,N)
            X(IVAR)=XIVAR
            EPSI=ZERO
            R1=DXMIN
            CALL MINSIM(X,N,NRED1,LIST,USERFN,R1,EPSI,NSTEPL,XR)
            IF(NSTEPL.LT.1) THEN
              NSTEP=-1
              GO TO 70
            END IF
C test for convergence
            IF(ABS(R1-FCONT).LT.EPS) THEN
              IF(ISIGN.LT.0) THEN
                DXMINS(IVAR)=X0(IVAR)-X(IVAR)
              ELSE
                DXPLUS(IVAR)=X(IVAR)-X0(IVAR)
              END IF
              GO TO 40
            ELSE
              IF(.NOT.LENCLS) THEN
C zero was not yet enclosed in last step
                IF(R1.GT.FCONT) THEN
C zero is now enclosed, perform first interval halving
                  LENCLS=.TRUE.
                  XBIG=X(IVAR)
                  X(IVAR)=HALF*(XBIG+XSMALL)
                ELSE
C zero not enclosed, widen range
                  DEL=DEL*TWO
                  X(IVAR)=XSMALL+SIGNUM*DEL
                END IF
              ELSE
C continue interval halving
                IF(R1.GT.FCONT) THEN
                  XBIG=X(IVAR)
                ELSE
                  XSMALL=X(IVAR)
                END IF
                X(IVAR)=HALF*(XBIG+XSMALL)
              END IF
              XIVAR=X(IVAR)
            END IF
   30     CONTINUE
          NSTEP=-2
          GO TO 70
   40     CONTINUE
   50   CONTINUE
C unfix variable IVAR
        LIST(IVAR)=1
   60 CONTINUE
   70 RETURN
      END
